northstar/src/runtime/process/mod.rs

```diff
@@ -272,87 +320,50 @@ impl super::state::Process for Process {
     }
 }
 
-/// Spawn a task that waits for the process to exit. Resolves to the exit status of `pid`.
-fn waitpid(
-    container: Container,
-    pid: Pid,
-    tx: EventTx,
-    stop: CancellationToken,
-) -> impl Future<Output = ExitStatus> {
-    let mut sigchld = signal::unix::signal(signal::unix::SignalKind::child())
-        .expect("Failed to set up signal handle for SIGCHLD");
-
-    task::spawn(async move {
-        // Check the status of the process after every SIGCHLD is received
-        let exit_status = loop {
-            sigchld.recv().await;
-            if let Some(exit) = exit_status(pid) {
-                stop.cancel();
-                break exit;
-            }
-        };
-
-        drop(
-            tx.send(Event::Container(
-                container,
-                ContainerEvent::Exit(exit_status.clone()),
-            ))
-            .await,
-        );
-        exit_status
-    })
-    .map(|r| r.expect("Task join error"))
-}
-
-/// Get exit status of process with `pid` or None
-fn exit_status(pid: Pid) -> Option<ExitStatus> {
+/// Get exit status of process with `pid` if the process already exited. Panic otherwise.
+fn init_exit_status(pid: Pid) -> ExitStatus {
```

Comment:
the `init_` part is making the function name confusing. The function is calling a blocking `waitpid` on a `pid` and panicking if something else to `exit` or `signaled` is returned. Maybe `reap_child` is a bit better name?

Alternatively, maybe it would be nicer to have and use a `async fn waitpid` similar the remove one:
```rust
async fn waitpid(pid: Pid) -> ExitStatus {
    let mut sigchld = signal::unix::signal(signal::unix::SignalKind::child()).expect("...");
    loop {
        sigchld.recv().await;
        if let Some(status) = exit_status(pid) {
            return status
        }
    }
}
```

Reply:
This is basically what we have had before. The drawback is that with the SIGCHLD we don't get the information about the pid that raised the SIGCHLD and we have to poll all running containers. As a tradeoff we call the (blocking) waitpid after we're sure that init exited because we either read the exit status of the child on the channel *or* the channel is closed.

Reply:
yeah, after thinking about it, it feels like the asynchronous `waitpid` is likely to suffer from race conditions with multiple child processes. But back to the original point, the name is a bit odd isn't it? how about a `waitpid` with some `block_in_place` inside?

Reply:
There's no possibility to `block_in_place` on a single threaded executor. See comment above.

What's your suggestion for the fns name? `reap_child` is not enough since the result of `waitpid` is used if there's not exit status received via the pipe. (sigkill case).

Reply:
`wait_exit_status(_blocking)`?

Reply:
That's why I originally named it `waitpid` ;-)

`wait_exit_status` is fine for me.

Reply:
well to be fair, `waitpid` was inlined and this is not the same :stuck_out_tongue_winking_eye: